packages/next/build/webpack/plugins/next-trace-entrypoints-plugin.ts

```diff
@@ -0,0 +1,253 @@
+                throw e
+              }
+            }
+            const readlink = (path: string) => {
+              try {
+                return compilation.inputFileSystem.readlinkSync(path)
+              } catch (e) {
+                if (
+                  e.code !== 'EINVAL' &&
+                  e.code !== 'ENOENT' &&
+                  e.code !== 'UNKNOWN'
+                ) {
+                  throw e
+                }
+                return null
+              }
+            }
+            const stat = (path: string) => {
+              try {
+                return compilation.inputFileSystem.statSync(path)
+              } catch (e) {
+                if (e.code === 'ENOENT') {
+                  return null
+                }
+                throw e
+              }
+            }
+
+            const nftCache = {}
+            const entryPaths = Array.from(entryModMap.keys())
+
+            for (const entry of entryPaths) {
+              depModMap.clear()
+              const entryMod = entryModMap.get(entry)
+              // TODO: investigate caching, will require ensuring no traced
+              // files in the cache have changed, we could potentially hash
+              // all traced files and only leverage the cache if the hashes
+              // match
+              // const cachedTraces = entryMod.buildInfo?.cachedNextEntryTrace
+
+              // Use cached trace if available and trace version matches
+              // if (
+              //   isWebpack5 &&
+              //   cachedTraces &&
+              //   cachedTraces.version === TRACE_OUTPUT_VERSION
+              // ) {
+              //   this.entryTraces.set(
+              //     entryNameMap.get(entry)!,
+              //     cachedTraces.tracedDeps
+              //   )
+              //   continue
+              // }
+              const collectDependencies = (mod: any) => {
+                if (!mod || !mod.dependencies) return
+
+                for (const dep of mod.dependencies) {
+                  const depMod = getModuleFromDependency(compilation, dep)
+
+                  if (depMod?.resource && !depModMap.get(depMod.resource)) {
+                    depModMap.set(depMod.resource, depMod)
+                    collectDependencies(depMod)
+                  }
+                }
+              }
+              collectDependencies(entryMod)
+
+              const toTrace: string[] = [entry, ...depModMap.keys()]
+
+              const root = nodePath.parse(process.cwd()).root
+              const result = await nodeFileTrace(toTrace, {
+                base: root,
+                cache: nftCache,
+                processCwd: this.appDir,
+                readFile,
+                readlink,
+                stat,
+                ignore: TRACE_IGNORES,
+              })
+
+              const tracedDeps: string[] = []
+
+              for (const file of result.fileList) {
+                if (result.reasons[file].type === 'initial') {
```

Comment:
Why ignore `initial` here?

Reply:
We don't want to include the source files themselves in the trace list